libs/pika/async_mpi/include/pika/async_mpi/trigger_mpi.hpp

```diff
@@ -0,0 +1,263 @@
+                                // we just assume the return from mpi_test is always MPI_SUCCESS
+                                ex::set_value(PIKA_MOVE(r.op_state.receiver));
+                                break;
+                            }
+                            case handler_mode::new_task:
+                            {
+                                // The callback will call set_value/set_error inside a new task
+                                // and execution will continue on that thread
+                                detail::schedule_task_callback(
+                                    request, PIKA_MOVE(r.op_state.receiver));
+                                break;
+                            }
+                            case handler_mode::continuation:
+                            {
+                                // The callback will call set_value/set_error
+                                // execution will continue on the callback thread
+                                detail::set_value_request_callback_void<>(request, r.op_state);
+                                break;
+                            }
+                            case handler_mode::suspend_resume:
+                            {
+                                // suspend is invalid except on a pika thread
+                                PIKA_ASSERT(pika::threads::detail::get_self_id());
+                                // the callback will resume _this_ thread
+                                std::unique_lock l{r.op_state.mutex_};
+                                resume_request_callback(request, r.op_state);
+                                if (use_HP_completion(r.op_state.mode_flags_))
+                                {
+                                    threads::detail::thread_data::scoped_thread_priority
+                                        set_restore(execution::thread_priority::high);
+                                    r.op_state.cond_var_.wait(
+                                        l, [&]() { return r.op_state.completed; });
+                                }
+                                else
+                                {
+                                    r.op_state.cond_var_.wait(
+                                        l, [&]() { return r.op_state.completed; });
+                                }
+                                // call set_value/set_error depending on mpi return status
+                                set_value_error_helper(
+                                    r.op_state.status, PIKA_MOVE(r.op_state.receiver));
+                                break;
+                            }
+                            default:
+                            {
+                                // I bet the "review" police don't like this exception
+                                throw mpi_exception(MPI_ERR_UNKNOWN, "eek!");
+                                break;
```

Comment:
:rotating_light:
Reaching this point is not an MPI problem, it's a logic error and we should terminate here.
```suggestion
                                PIKA_UNREACHABLE;
```
:rotating_light: